### PR TITLE
Refactor Determined profiler into Core API [MD-10, MD-302]

### PR DESCRIPTION
**Pull request by azhou-determined**
## Checklist

- [ ] Changes have been manually QA'd
- [ ] User-facing API changes need the "User-facing API Change" label.
- [ ] Release notes should be added as a separate file under `docs/release-notes/`.
  See [Release Note](https://github.com/determined-ai/determined/blob/master/docs/release-notes/README.md) for details.
- [ ] Licenses should be included for new code which was copied and/or modified from any external code.

### CODE REVIEW

File: harness/determined/core/_profiler.py [hunk 1]
```diff
@@ -0,0 +1,396 @@
...
+        agent_id: str,
+        trial_id: int,
+        run_id: int,
+        distributed: core.DistributedContext,
+    ) -> None:
+        self._session = session
+        self._agent_id = agent_id
+        self._trial_id = trial_id
+        self._run_id = run_id
+        # xxx: pass in train context here to use for metrics reporting API?
```
Comment (rb-determined-ai) [inline harness/determined/core/_profiler.py:37]: that seems acceptable to me... you are already passing the distributed context as a dependency.  Feels morally equivalent.

File: harness/determined/core/_profiler.py [hunk 2]
```diff
@@ -0,0 +1,396 @@
...
+        self._collector.start()
+        self._shipper.start()
+        self._on = True
+
+    def off(self) -> None:
+        if not self._on:
+            logger.warning(f"Profiler is already off.")
+            return
+        logger.info(f"Stopping system metrics profiling.")
+        self.close()
```
Comment (rb-determined-ai) [inline harness/determined/core/_profiler.py:78]: pick one; either use `on/off` and let it be a lightswitch that can be flipped back and forth, or let `close()` invalidate the resource and don't let it be reopened.

Practically anything with `.close()` is unusable after you call `.close()`.

File: harness/determined/core/_profiler.py [hunk 3]
```diff
@@ -0,0 +1,396 @@
...
+        self._distributed = distributed
+        self._on = False
+
+        self._collector = None
+        self._shipper = None
+
+    # xxx: other potential configs (collection interval, flush interval, max collection time)
+    def on(self, interval: int = 1) -> None:
+        if self._on:
+            logger.warning(f"Profiler is already on.")
```
Comment (rb-determined-ai) [inline harness/determined/core/_profiler.py:47]: don't warn, just document that it's idempotent.

Same with off.

File: harness/determined/core/_profiler.py [hunk 4]
```diff
@@ -0,0 +1,396 @@
...
+            disk_usage = psutil.disk_usage(path)
+            metrics.update({path: {"disk_util": disk_usage.percent}})
+        return metrics
+
+
+class _Memory(_MetricGroupCollector):
+    group = "memory"
+
+    def collect(self) -> Dict[str, Any]:
+        free_mem_bytes = psutil.virtual_memory().available
```
Comment (rb-determined-ai) [inline harness/determined/core/_profiler.py:188]: emit more than just free memory; free memory is [not a very complete picture](https://www.linuxatemyram.com/) of memory usage.

File: harness/determined/core/_profiler.py [hunk 5]
```diff
@@ -0,0 +1,396 @@
...
+    def __init__(self):
+        super().__init__()
+
+        # xxx: map[uuid] -> handle
+        self._pynvml_device_handles: Dict[str, Any] = {}
+
+        if pynvml:
+            self._init_pynvml()
+        else:
+            logging.warning(f"pynvml module not found. GPU metrics will not be collected.")
```
Comment (rb-determined-ai) [inline harness/determined/core/_profiler.py:234]: should not be an `f`string

File: harness/determined/core/_profiler.py [hunk 6]
```diff
@@ -0,0 +1,396 @@
...
+
+    def close(self) -> None:
+        pass
+
+
+class _MetricGroupCollector(metaclass=abc.ABCMeta):
+    group: str
+
+    def __init__(self):
+        pass
```
Comment (rb-determined-ai) [inline harness/determined/core/_profiler.py:108]: don't define pointless methods

File: harness/determined/core/_profiler.py [hunk 7]
```diff
@@ -0,0 +1,396 @@
...
+    @abc.abstractmethod
+    def collect(self) -> Dict[str, Any]:
+        pass
+
+
+class _Network(_MetricGroupCollector):
+    group = "network"
+
+    def __init__(self) -> None:
+        super().__init__()
```
Comment (rb-determined-ai) [inline harness/determined/core/_profiler.py:119]: don't call pointless methods

File: harness/determined/core/_profiler.py [hunk 8]
```diff
@@ -0,0 +1,396 @@
...
+        self._metric_collectors = [
+            _GPU(),
+            _CPU(),
+            _Memory(),
+            _Disk(),
+            _Network(),
+        ]
+        self._should_exit = threading.Event()
+
+        super().__init__(daemon=True)
```
Comment (rb-determined-ai) [inline harness/determined/core/_profiler.py:307]: Why is this daemon=True?

File: harness/determined/core/_profiler.py [hunk 9]
```diff
@@ -0,0 +1,396 @@
...
+
+    def run(self) -> None:
+        collection_stop_ts = time.time() + self._max_collection_s
+        while not self._should_exit.is_set():
+            if time.time() > collection_stop_ts:
+                # Max collection time reached. Notify queue consumers and exit.
+                self._metrics_queue.put(_ShutdownMessage())
```
Comment (rb-determined-ai) [inline harness/determined/core/_profiler.py:314]: always put your shutdown sentinel in a `finally` block.  As written, if you throw an exception your consumer will never shut down.

File: harness/determined/core/_profiler.py [hunk 10]
```diff
@@ -0,0 +1,396 @@
...
+        metrics: Dict[str, Any],
+        timestamp: datetime,
+    ):
+        self.group = group
+        self.metrics = metrics
+        self.timestamp = timestamp
+
+
+class _ShutdownMessage:
+    pass
```
Comment (rb-determined-ai) [inline harness/determined/core/_profiler.py:281]: just my opinion, but `None` is a fine sentinel for "this queue has nothing left and you should die".

File: harness/determined/core/_profiler.py [hunk 11]
```diff
@@ -0,0 +1,396 @@
...
+                # Max collection time reached. Notify queue consumers and exit.
+                self._metrics_queue.put(_ShutdownMessage())
+                return
+
+            self._collect_metrics()
+            self._should_exit.wait(self._collection_interval)
+
+    def stop(self) -> None:
+        self._should_exit.set()
+        self._metrics_queue.put(_ShutdownMessage())
```
Comment (rb-determined-ai) [inline harness/determined/core/_profiler.py:322]: Find a way not to need this "doubletap" cleanup step.

File: harness/determined/core/_profiler.py [hunk 12]
```diff
@@ -0,0 +1,396 @@
...
+        self._metrics: List[_Metric] = []
+
+        super().__init__(daemon=True)
+
+    def run(self) -> None:
+        while True:
+            next_flush_ts = time.time() + self._flush_interval
+            # Pop messages from queue until the next scheduled flush time.
+            while time.time() < next_flush_ts:
+                deadline = next_flush_ts - time.time()
```
Comment (rb-determined-ai) [inline harness/determined/core/_profiler.py:361]: whoa, you got a really sneaky bug here.

Your concept of "now" is changing because each call to `time.time()` is going to change.  So you can end up with `deadline` as a negative value, and `queue.get()` will block forever.

Comment (rb-determined-ai) [reply harness/determined/core/_profiler.py:361]: I like to set a variable called `now` and reuse it during my time calculation.  Example [here](https://github.com/determined-ai/determined/blob/2dfc4f208dfb1a3afa8ee3c4de14a71216934493/master/static/srv/ship_logs.py#L311-L321).

File: harness/determined/core/_profiler.py [hunk 13]
```diff
@@ -0,0 +1,396 @@
...
+                    # Timeout reached, no messages in queue.
+                    break
+
+            self._ship()
+
+    def stop(self) -> None:
+        self._metrics_queue.put(_ShutdownMessage())
+
+    def _ship(self) -> None:
+        # xxx: change this to popleft from list?
```
Comment (rb-determined-ai) [inline harness/determined/core/_profiler.py:380]: or, better idea: let metrics be an input to ship: `def _ship(self, metrics)`

That way it doesn't need to be on `self` at all, and you can set `metrics = []` at the top of the `while True` loop, and the lifetime of the metrics array is a lot clearer.

File: harness/determined/core/_profiler.py [hunk 14]
```diff
@@ -0,0 +1,396 @@
...
+        for metric in self._metrics:
+            # Append agent ID to every metric reported.
+            metrics = {self._agent_id: metric.metrics}
+            self._report_metrics(group=metric.group, timestamp=metric.timestamp, metrics=metrics)
```
Comment (rb-determined-ai) [inline harness/determined/core/_profiler.py:384]: wut?  if you are doing a separate HTTP request for each metric, why are you even bothering with a loop to collect batches of metrics?

File: harness/determined/keras/callbacks.py [hunk 15]
```diff
@@ -711,33 +711,3 @@ def on_train_workload_begin(
     def on_train_workload_end(self, total_batches_trained: int, logs: Dict) -> None:
         tf.keras.callbacks.TensorBoard.on_epoch_end(self, self.workload_end_count, logs)
         self.workload_end_count += 1
-
-
-class _DeterminedProfiler(Callback):
```
Comment (rb-determined-ai) [inline harness/determined/keras/callbacks.py:716]: :tada:

File: harness/determined/pytorch/_trainer.py [hunk 16]
```diff
@@ -62,23 +61,8 @@ def configure_profiler(
...
-            agent_id=self._info.agent_id,
-            master_url=self._info.master_url,
-            profiling_is_enabled=enabled,
-            global_rank=self._core.distributed.get_rank(),
-            local_rank=self._core.distributed.get_local_rank(),
-            begin_on_batch=begin_on_batch,
-            end_after_batch=end_after_batch,
-            sync_timings=sync_timings,
-        )
+        # xxx: deprecate this method
```
Comment (rb-determined-ai) [inline harness/determined/pytorch/_trainer.py:64]: yes, and emit a warning.

**Review by rb-determined-ai — APPROVED** [review]
looks fantastic.  Very easy to review this code, nice work.

**Comment by azhou-determined** [comment]
merged in main repo a long time ago
